### PR TITLE
feat: create number field as base and MSFT component

### PR DESCRIPTION
**Pull request by nalogiudice**
Closes [#1365](https://github.com/Microsoft/fast-dna/issues/1365)
# Description
Basic `<input type="number" />` component.

## Issue type checklist

- [ ] **Chore**: A change that does not impact distributed packages.
- [ ] **Bug fix**: A change that fixes an issue, link to the issue above.
- [x] **New feature**: A change that adds functionality.

## Process & policy checklist

- [x] I have added tests for my changes.
- [x] I have tested my changes.
- [x] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/Microsoft/fast-dna/blob/master/CONTRIBUTING.md) documentation and followed the [standards](https://microsoft.github.io/fast-dna/docs/en/contributing/standards) for this project.

### CODE REVIEW

File: packages/fast-components-react-base/src/number-field/number-field.tsx [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+            attributes.readOnly = true;
+        }
+
+        if (typeof this.props.required === "boolean") {
+            attributes.required = true;
+        }
+
+        if (typeof this.props.step === "number") {
+            attributes.step = this.props.step;
+        }
```
Comment (janechu) [inline packages/fast-components-react-base/src/number-field/number-field.tsx]: I question the need to enumerate these, they should already be specified by the unhandledProps, since this should be typed as an HTMLInputElement instead of HTMLDivElement.

File: packages/fast-components-react-base/src/number-field/number-field.schema.json [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+            "type": "boolean"
+        },
+        "readOnly": {
+            "title": "HTML readOnly attribute",
+            "type": "boolean"
+        },
+        "required": {
+            "title": "HTML required attribute",
+            "type": "boolean"
+        }
```
Comment (janechu) [inline packages/fast-components-react-base/src/number-field/number-field.schema.json]: please also include "value"

File: packages/fast-components-react-base/src/number-field/number-field.tsx [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+            attributes.disabled = true;
+        }
+
+        if (typeof this.props.max === "number") {
+            attributes.max = this.props.max;
+        }
+
+        if (typeof this.props.min === "number") {
+            attributes.min = this.props.min;
+        }
```
Comment (eljefe223) [inline packages/fast-components-react-base/src/number-field/number-field.tsx]: Should this be able to handle negative values? It currently doesn't. I can manually type negative values in the input but cannot in the Configuration panel

File: packages/fast-components-styles-msft/src/utilities/fields.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+            cursor: "not-allowed",
+            "&::placeholder": {
+                ...typographyPattern.disabled,
+            },
+        },
+        "&::placeholder": {
+            color: foregroundNormal,
+        },
+    };
+}
```
Comment (chrisdholt) [inline packages/fast-components-styles-msft/src/utilities/fields.ts]: Please break this refactor out into a different PR. Changing text field and text area don't align with this feature's intent.

File: packages/fast-components-react-base/src/number-field/number-field.tsx (hunk 3 above)
Comment (nalogiudice) [reply packages/fast-components-react-base/src/number-field/number-field.tsx]: It should... I will look into it.

File: packages/fast-components-react-base/src/number-field/number-field.tsx (hunk 1 above)
Comment (nalogiudice) [reply packages/fast-components-react-base/src/number-field/number-field.tsx]: `HTMLDivElement` to `HTMLInputElement` sure - I missed that.

> I question the need to enumerate **these**

What are these specifically @janechu ? My thought was to expose the common attributes users would expect with a number field. Or else you get a configuration pane without much configuration.

File: packages/fast-components-styles-msft/src/utilities/fields.ts (hunk 4 above)
Comment (nalogiudice) [reply packages/fast-components-styles-msft/src/utilities/fields.ts]: Code climate would have complained if I had repeated the same pattern for a third time. I suppose I could pull this out into a small PR to get in first

File: packages/fast-components-react-base/src/number-field/number-field.schema.json (hunk 2 above)
Comment (nalogiudice) [reply packages/fast-components-react-base/src/number-field/number-field.schema.json]: Sure

File: packages/fast-components-styles-msft/src/utilities/fields.ts (hunk 4 above)
Comment (nalogiudice) [reply packages/fast-components-styles-msft/src/utilities/fields.ts]: @chrisdholt pulled out into another [PR](https://github.com/Microsoft/fast-dna/pull/1375)

Comment (chrisdholt) [reply packages/fast-components-styles-msft/src/utilities/fields.ts]: Thanks - Unrelated to code climate, this is just all around different work than the purpose of this PR. I'll take a look over there.

File: packages/fast-components-react-base/src/number-field/number-field.tsx (hunk 3 above)
Comment (nalogiudice) [reply packages/fast-components-react-base/src/number-field/number-field.tsx]: Everything seems to be working as expected. I can set a negative min in the configurator and it functions as expected. Can you look into your issue again?

File: packages/fast-components-react-base/src/number-field/number-field.tsx (hunk 1 above)
Comment (janechu) [reply packages/fast-components-react-base/src/number-field/number-field.tsx]: All of the props, the fact that this is typed as an HTMLInputElement means this should not need you to enumerate disabled, step, min, max etc.

Comment (janechu) [reply packages/fast-components-react-base/src/number-field/number-field.tsx]: I think it's fine to leave them in the JSON schema because we want to be able to configure those values in the form generator, but you shouldn't need these in the handledProps

File: packages/fast-components-react-base/src/number-field/number-field.tsx (hunk 3 above)
Comment (eljefe223) [reply packages/fast-components-react-base/src/number-field/number-field.tsx]: I still can't set a negative value in the configuration panel... Its probably not a blocker or even an issue with your code. Probably an issue in fast form

Comment (nalogiudice) [reply packages/fast-components-react-base/src/number-field/number-field.tsx]: What browser are you using? I am able to set negative values in the form generator no problem:

screen shot 2019-02-18 at 3 48 43 pm

Comment (eljefe223) [reply packages/fast-components-react-base/src/number-field/number-field.tsx]: So weird... I'm using Chrome

Comment (nalogiudice) [reply packages/fast-components-react-base/src/number-field/number-field.tsx]: Anyone else seeing this?

File: packages/fast-components-react-msft/src/number-field/examples.data.tsx [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        },
+        {
+            required: true,
+        },
+        {
+            placeholder: "Placeholder",
+        },
+        {
+            name: "Name",
+        },
```
Comment (chrisdholt) [inline packages/fast-components-react-msft/src/number-field/examples.data.tsx]: can we add an example which includes `value`?

File: packages/fast-components-react-base/src/number-field/number-field.tsx [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    protected generateClassNames(): string {
+        return super.generateClassNames(get(this.props, "managedClasses.numberField"));
+    }
+
+    /**
+     * Generates attributes based on props
+     */
+    private assignAttributes(): Partial<NumberFieldUnhandledProps> {
+        if (typeof this.props.value === "number") {
+            return { defaultValue: `${this.props.value}` };
```
Comment (janechu) [inline packages/fast-components-react-base/src/number-field/number-field.tsx]: I would not re-assign value to defaultValue, these are two separate ideas, one which results in a controlled component and one which results in uncontrolled, see: https://reactjs.org/docs/uncontrolled-components.html#default-values

I don't think it's necessary to specify either of these in the component or interface, only in the JSON schema.

File: packages/fast-components-react-base/src/number-field/number-field.tsx [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    NumberFieldClassNameContract,
+} from "@microsoft/fast-components-class-name-contracts-base";
+import { TextFieldType } from "../text-field/index";
+
+class NumberField extends Foundation<
+    NumberFieldHandledProps,
+    NumberFieldUnhandledProps,
+    {}
+> {
+    public static defaultProps: Partial<NumberFieldProps> = {};
```
Comment (chrisdholt) [inline packages/fast-components-react-base/src/number-field/number-field.tsx]: If we aren't using this we should be able to remove this.

File: packages/fast-components-react-base/src/number-field/number-field.props.ts [hunk 8]
```diff
@@ -0,0 +1,18 @@
+import * as React from "react";
+import {
+    ManagedClasses,
+    NumberFieldClassNameContract,
+} from "@microsoft/fast-components-class-name-contracts-base";
+
+export interface NumberFieldManagedClasses
+    extends ManagedClasses<NumberFieldClassNameContract> {}
+export interface NumberFieldUnhandledProps
+    extends React.HTMLAttributes<HTMLInputElement> {}
```
Comment (chrisdholt) [inline packages/fast-components-react-base/src/number-field/number-field.props.ts]: @nalogiudice we should remove "type" here using utility-types

File: packages/fast-components-react-base/src/number-field/number-field.tsx [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+    public static displayName: string = "NumberField";
+
+    protected handledProps: HandledProps<NumberFieldHandledProps> = {
+        managedClasses: void 0,
+        value: void 0,
```
Comment (chrisdholt) [inline packages/fast-components-react-base/src/number-field/number-field.tsx]: @janechu and @nalogiudice can we jump on discord and sync up sometime today about the handled prop contract here?

File: packages/fast-components-react-base/src/number-field/number-field.tsx (hunk 6 above)
Comment (nalogiudice) [reply packages/fast-components-react-base/src/number-field/number-field.tsx]: I believe I have resolved this.

File: packages/fast-components-react-base/src/number-field/number-field.props.ts (hunk 8 above)
Comment (nalogiudice) [reply packages/fast-components-react-base/src/number-field/number-field.props.ts]: Yep good catch

File: packages/fast-components-react-base/src/number-field/number-field.tsx (hunk 9 above)
Comment (nalogiudice) [reply packages/fast-components-react-base/src/number-field/number-field.tsx]: I think it will only be `managedClasses`. @janechu has specified all props can be passed via unhandled props. If we need a chat that's fine, or you may just be able to talk to Jane.

File: packages/fast-components-react-msft/src/number-field/examples.data.tsx (hunk 5 above)
Comment (nalogiudice) [reply packages/fast-components-react-msft/src/number-field/examples.data.tsx]: Added

File: packages/fast-components-react-base/src/number-field/number-field.tsx (hunk 3 above)
Comment (nalogiudice) [reply packages/fast-components-react-base/src/number-field/number-field.tsx]: Going to resolve for now, but if someone else runs into this issue, I will reopen.
